pkg/controller/metrics/metricsWithDynamicLabels.go

```diff
@@ -0,0 +1,91 @@
+type metricsWithDyanamicLabels interface {
+	LogMetricWithDynamicLabels(metav1.Object, log.FieldLogger, float64)
+}
+
+type CounterWithDynamicLabels struct {
+	*prometheus.CounterVec
+	FixedLabels    map[string]string
+	OptionalLabels map[string]string
+}
+
+// LogMetricWithDynamicLabels logs the counter metric. It simply increments the value, so last parameter isn't used.
+func (c CounterWithDynamicLabels) LogMetricWithDynamicLabels(obj metav1.Object, log log.FieldLogger, _ float64) {
+	curriedVec, err := c.CounterVec.CurryWith(c.FixedLabels)
+	// Maps cannot be sorted, however we would want to loop through them in a sorted fashion to ensure all the metric
+	// labels are added in an expected order. Get the slice of map keys first, sort them and then use it to loop over
+	// the map
+	for _, v := range GetSortedKeys(c.OptionalLabels) {
+		curriedVec, err = curriedVec.CurryWith(prometheus.Labels{v: GetLabelValue(obj, c.OptionalLabels[v])})
+	}
+	if err != nil {
+		// The possibility of hitting an error here is low, but report it and don't observe the metric in case it happens
+		log.WithError(err).Error("error observing metric")
+	} else {
+		curriedVec.WithLabelValues().Inc()
+	}
+}
+
+type HistogramWithDynamicLabels struct {
+	*prometheus.HistogramVec
+	FixedLabels    map[string]string
+	OptionalLabels map[string]string
+}
+
+// LogMetricWithDynamicLabels observes the histogram metric with val value.
+func (h HistogramWithDynamicLabels) LogMetricWithDynamicLabels(obj metav1.Object, log log.FieldLogger, val float64) {
+	curriedVec, err := h.HistogramVec.CurryWith(h.FixedLabels)
+	// Maps cannot be sorted, however we would want to loop through them in a sorted fashion to ensure all the metric
+	// labels are added in an expected order. Get the slice of map keys first, sort them and then use it to loop over
+	// the map
+	for _, v := range GetSortedKeys(h.OptionalLabels) {
+		curriedVec, err = curriedVec.CurryWith(prometheus.Labels{v: GetLabelValue(obj, h.OptionalLabels[v])})
+	}
+	if err != nil {
+		// The possibility of hitting an error here is low, but report it and don't observe the metric in case it happens
+		log.WithError(err).Error("error observing metric")
+	} else {
+		curriedVec.WithLabelValues().Observe(val)
+	}
+}
+
+var _ metricsWithDyanamicLabels = CounterWithDynamicLabels{}
+var _ metricsWithDyanamicLabels = HistogramWithDynamicLabels{}
+
+// GetOptionalClusterTypeLabels reads the AdditionalClusterDeploymentLabels from the metrics config and returns the same
+// as a map
+func GetOptionalClusterTypeLabels(mConfig *metricsconfig.MetricsConfig) map[string]string {
+	var mapClusterTypeLabelToValue = make(map[string]string)
+	// Currently "cluster_type" is a non-optional label for these metrics.
+	// TODO: remove as a part of https://issues.redhat.com/browse/HIVE-2077
+	if _, ok := mConfig.AdditionalClusterDeploymentLabels["cluster_type"]; !ok {
+		mapClusterTypeLabelToValue["cluster_type"] = hivev1.HiveClusterTypeLabel
+	}
```

Comment:
Hm.

I think I would rather add this default *only* if `AdditionalClusterDeploymentLabels` is not configured at all. E.g. what if the consumer wants to use that CD label, but name the metric label something different? They would end up with a duplicate. If they're using the new config field at all, we're not worried about backward compatibility: they can specify this explicitly.

And at that point I'm not even sure if we need to bother with HIVE-2077.

Reply:
Exactly.

I'm not worried about breaking dashboards. The dashboards, alerts, and HiveConfig are all defined in app-interface, so SD will, in one MR, either configure `cluster_type` explicitly with that exact label; or configure it with some other name and change the dashboards and alerts at the same time (unlikely).

So I'm standing by this comment.

Reply:
Okay, I've changed it accordingly